### PR TITLE
Add graph with efficient border edge / internal edge scan

### PR DESCRIPTION
**Pull request by TheMarex**
# Issue

Addresses #3779 by implementing a graph data structure based on a sorted StaticGraph.

## Tasklist
 - [x] add regression / cucumber cases (see docs/testing.md)
 - [x] review
 - [x] adjust for comments

## Requirements / Relations
Based on #3765 
Blocked by implicit edge sorting https://github.com/Project-OSRM/osrm-backend/issues/3737

### CODE REVIEW

**Comment by TheMarex** [comment]
We could even optimize the `if(IsBoundaryArc)` check away by storing an `uint8_t` per level for every boundary node that encodes how many border edges a node on each level.

Potential cost for a 4-level partition would be (4 * 8bit) = 32bit per border node.

Will benchmark/work further on this once we have the updater code in place.

File: include/util/static_graph.hpp [hunk 1]
```diff
@@ -96,39 +107,12 @@ template <typename EdgeDataT, bool UseSharedMemory = false> class StaticGraph
         return irange(BeginEdges(node), EndEdges(node));
     }
 
-    template <typename ContainerT> StaticGraph(const int nodes, const ContainerT &graph)
+    template <typename ContainerT> StaticGraph(const int nodes, const ContainerT &edges)
```
Comment (oxidase) [inline include/util/static_graph.hpp]: Here default constructor is mssing

File: include/partition/multi_level_graph.hpp [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    auto SortEdgesByHighestLevel(const std::vector<LevelID> &highest_border_level,
+                                 const ContainerT &edges) const
+    {
+        std::vector<std::uint32_t> permutation(edges.size());
+        std::iota(permutation.begin(), permutation.end(), 0);
+        std::sort(permutation.begin(),
+                  permutation.end(),
+                  [&highest_border_level](const auto &lhs, const auto &rhs) {
+                      // sort by source node and then by level in _decreasing_ order
+                      return std::tie(lhs.source, highest_border_level[rhs]) <
```
Comment (oxidase) [inline include/partition/multi_level_graph.hpp]: Here must be `edges[lhs].source` and `edges[rhs].source`

File: include/partition/multi_level_graph.hpp [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+  public:
+    template <typename ContainerT>
+    MultiLevelGraph(const MultiLevelPartition &mlp,
+                    const std::uint32_t num_nodes,
+                    const ContainerT &edges)
+    {
+        auto highest_border_level = GetHighestBorderLevel(mlp, edges);
+        auto permutation = SortEdgesByHighestLevel(highest_border_level, edges);
+        InitializeGraphFromSortedEdges(
```
Comment (oxidase) [inline include/partition/multi_level_graph.hpp]: `SuperT::InitializeFromSortedEdgeRange`

**Comment by TheMarex** [comment]
@oxidase thanks for the fixes! Pulled in your fixes and rebased.

**Comment by TheMarex** [comment]
Measurements from the Bayern instance on the benchmark query `/route/v1/driving/11.578946,48.138177;9.981766,53.547244` with 100 queries.

Median on this branch: `12.81ms`
Median on `master`: `13.31ms`

Less exciting then I hoped for but it's a start.

Adds about `0.5` seconds to the customization to construct the graph. In the future we can off-load this work to `osrm-partition`, but this requires some changes around how we store the edge-based-graph.

**Comment by TheMarex** [comment]
@oxidase can you have a look at this again? Some core parts changed and this is now hooked up the search code. We could also use this in the customization code for scans restricted to cells (`GetInternalEdgeRange`).

File: include/partition/multi_level_graph.hpp [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+    auto SortEdgesByHighestLevel(const std::vector<LevelID> &highest_border_level,
+                                 const ContainerT &edges) const
+    {
+        std::vector<std::uint32_t> permutation(edges.size());
+        std::iota(permutation.begin(), permutation.end(), 0);
+        tbb::parallel_sort(
+            permutation.begin(),
+            permutation.end(),
+            [&edges, &highest_border_level](const auto &lhs, const auto &rhs) {
+                // sort by source node and then by level in acending order
```
Comment (oxidase) [inline include/partition/multi_level_graph.hpp]: s/acending/ascending/

File: include/partition/multi_level_graph.hpp [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+    {
+        auto begin = SuperT::BeginEdges(node);
+        auto end = SuperT::BeginEdges(node) + node_to_edge_offset[node + level];
+        return util::irange<EdgeID>(begin, end);
+    }
+
+    EdgeID BeginBorderEdges(const LevelID level, const NodeID node) const
+    {
+        auto index = node * GetNumberOfLevels();
+        if (index >= node_to_edge_offset.size() - 1)
```
Comment (oxidase) [inline include/partition/multi_level_graph.hpp]: Don't understand this, can it be an assertion and 
`return SuperT::BeginEdges(node) + node_to_edge_offset[index + level];`?

File: include/partition/multi_level_graph.hpp [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+        auto end = SuperT::EndEdges(node);
+        return util::irange<EdgeID>(begin, end);
+    }
+
+    // Fast scan over all relevant internal edges, that is edges that will not
+    // leave the cell of that node at the given level
+    auto GetInternalEdgeRange(const LevelID level, const NodeID node) const
+    {
+        auto begin = SuperT::BeginEdges(node);
+        auto end = SuperT::BeginEdges(node) + node_to_edge_offset[node + level];
```
Comment (oxidase) [inline include/partition/multi_level_graph.hpp]: Here `node_to_edge_offset[index + level];`?

File: include/partition/multi_level_graph.hpp [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+    EdgeID BeginBorderEdges(const LevelID level, const NodeID node) const
+    {
+        auto index = node * GetNumberOfLevels();
+        if (index >= node_to_edge_offset.size() - 1)
+            return SuperT::BeginEdges(node);
+        else
+            return SuperT::BeginEdges(node) + node_to_edge_offset[index + level];
+    }
+
+    // We save the level as senitel at the end
```
Comment (oxidase) [inline include/partition/multi_level_graph.hpp]: s/senitel/sentinel/

File: include/partition/multi_level_graph.hpp [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+    auto GetBorderEdgeRange(const LevelID level, const NodeID node) const
+    {
+        auto begin = BeginBorderEdges(level, node);
+        auto end = SuperT::EndEdges(node);
+        return util::irange<EdgeID>(begin, end);
+    }
+
+    // Fast scan over all relevant internal edges, that is edges that will not
+    // leave the cell of that node at the given level
+    auto GetInternalEdgeRange(const LevelID level, const NodeID node) const
```
Comment (oxidase) [inline include/partition/multi_level_graph.hpp]: There is no test coverage for `GetInternalEdgeRange`.
Could you also add a test check that `GetInternalEdgeRange(level, node) ∪ GetBorderEdgeRange(level, node) = GetAdjacentEdgeRange(node)` for all nodes and all levels in the unit test graph?

File: include/partition/multi_level_graph.hpp (hunk 5 above)
Comment (TheMarex) [inline include/partition/multi_level_graph.hpp]: `node_to_edge` will have the size defined by the maximum ID of border nodes. Right now that maximum ID is somewhere around `|V|` but if we implement a sorting of NodeIDs based on boundary level, we will save a lot of memory here (> 80%, typically less then 20% of the nodes are boundary nodes).

File: include/partition/multi_level_graph.hpp (hunk 6 above)
Comment (TheMarex) [inline include/partition/multi_level_graph.hpp]: Good catch. 👍

File: include/partition/multi_level_graph.hpp (hunk 8 above)
Comment (TheMarex) [inline include/partition/multi_level_graph.hpp]: Yep added test coverage for that.

**Comment by TheMarex** [comment]
@oxidase addressed the comments and integrated the`GetInternalEdgeRange` in the cell customizer. On Bayern I could sadly not measure an improvement.

**Comment by oxidase** [comment]
Munich->Würzburg comparison shows 0.6ms improvement for ~10ms query times
compare1

LGTM! Please add a TODO comment in `BeginBorderEdges` for `return SuperT::BeginEdges(node);` that it is planned for further improvements or even  better open a new issue with the code snippet and remove it because there no current use of it.

**Pull request merged**